src/Interpreters/Context.cpp

```diff
@@ -1394,59 +1396,64 @@ StoragePtr Context::executeTableFunction(const ASTPtr & table_expression, const
+                        if (asterisk)
+                            throw Exception(ErrorCodes::ILLEGAL_COLUMN, "Asterisk cannot be mixed with column list in INSERT SELECT query.");
 
-                /// Check that all identifiers are column names from insertion table and not virtual column names from storage.
-                for (const auto & column_name : columns_names)
-                {
-                    if (!structure_hint.has(column_name) || virtual_column_names.contains(column_name))
-                    {
-                        use_columns_from_insert_query = false;
-                        break;
+                        structure_hint.add({ identifier->name(), insert_column->type });
                     }
+
+                    if (asterisk)
+                        --insert_structure_end;
+                    else
+                        ++insert_column;
                 }
+                else if ((*expression)->as<ASTAsterisk>())
+                {
+                    if (asterisk)
+                        throw Exception(ErrorCodes::ILLEGAL_COLUMN, "Only one asterisk can be used in INSERT SELECT query.");
+                    if (!structure_hint.empty())
+                        throw Exception(ErrorCodes::ILLEGAL_COLUMN, "Asterisk cannot be mixed with column list in INSERT SELECT query.");
 
-                /// If we don't have asterisk but only subset of columns, we should use
-                /// structure from insertion table only in case when table function
-                /// supports reading subset of columns from data.
-                if (use_columns_from_insert_query && !have_asterisk && !columns_names.empty())
+                    asterisk = true;
+                }
+                else
                 {
-                    /// For input function we should check if input format supports reading subset of columns.
-                    if (table_function_ptr->getName() == "input")
-                        use_columns_from_insert_query = FormatFactory::instance().checkIfFormatSupportsSubsetOfColumns(getInsertFormat());
+                    if (asterisk)
+                        --insert_structure_end;
                     else
-                        use_columns_from_insert_query = table_function_ptr->supportsReadingSubsetOfColumns();
+                        ++insert_column;
                 }
             }
 
-            if (use_columns_from_insert_query)
+            if (expression != expression_list.end())
+                throw Exception(ErrorCodes::NUMBER_OF_COLUMNS_DOESNT_MATCH, "Number of columns in insert table less than required by SELECT expression.");
+
+            if (asterisk)
+            {
+                for (; insert_column != insert_structure_end; ++insert_column)
+                    structure_hint.add({ insert_column->name, insert_column->type });
+            }
+
+            if (!structure_hint.empty())
                 table_function_ptr->setStructureHint(structure_hint);
```

Comment:
I would prefer not to set a hint when we have a list of column names and format that doesn't support reading subset of columns.
For example if we have `data.csv` with 4 columns but we want to insert into table only two of them:
```
insert into test select c1, c2 from file(data.csv)
```
And it won't work now

Reply:
so, we want `use_structure_from_insertion_table_in_table_functions=auto`?